core/src/main/java/org/elasticsearch/script/CachingCompiler.java

```diff
@@ -0,0 +1,385 @@
+        }
+        return cacheKeyFromClusterState(source);
+    }
+
+    /**
+     * Check that a script compiles before attempting to store it.
+     */
+    public final void checkCompileBeforeStore(StoredScriptSource source) {
+        CacheKeyT cacheKey = cacheKeyFromClusterState(source);
+        try {
+            if (areAnyScriptContextsEnabled(cacheKey, ScriptType.STORED)) {
+                Object compiled = compile(ScriptType.STORED, cacheKey);
+
+                if (compiled == null) {
+                    throw new IllegalArgumentException("failed to parse/compile");
+                }
+            } else {
+                throw new IllegalArgumentException("cannot be run under any context");
+            }
+        } catch (ScriptException good) {
+            throw good;
+        } catch (Exception exception) {
+            throw new IllegalArgumentException("failed to parse/compile", exception);
+        }
+    }
+
+    @Override
+    public final void clusterChanged(ClusterChangedEvent event) {
+        clusterState = event.state();
+    }
+
+    /**
+     * Listener to manage metrics for the script cache.
+     */
+    private class CacheRemovalListener implements RemovalListener<CacheKeyT, CompiledScript> {
+        @Override
+        public void onRemoval(RemovalNotification<CacheKeyT, CompiledScript> notification) {
+            if (logger.isDebugEnabled()) {
+                logger.debug("removed {} from cache, reason: {}", notification.getValue(),
+                        notification.getRemovalReason());
+            }
+            scriptMetrics.onCacheEviction();
+        }
+    }
+
+    private class DirectoryChangesListener implements FileChangesListener {
+        private Tuple<String, String> getScriptNameExt(Path file) {
+            Path scriptPath = scriptsDirectory.relativize(file);
+            int extIndex = scriptPath.toString().lastIndexOf('.');
+            if (extIndex <= 0) {
+                return null;
+            }
+
+            String ext = scriptPath.toString().substring(extIndex + 1);
+            if (ext.isEmpty()) {
+                return null;
+            }
+
+            String scriptName = scriptPath.toString().substring(0, extIndex)
+                    .replace(scriptPath.getFileSystem().getSeparator(), "_");
+            return new Tuple<>(scriptName, ext);
+        }
+
+        @Override
+        public void onFileInit(Path file) {
+            Tuple<String, String> scriptNameExt = getScriptNameExt(file);
+            if (scriptNameExt == null) {
+                logger.debug("Skipped script with invalid extension : [{}]", file);
+                return;
+            }
+            if (logger.isTraceEnabled()) {
+                logger.trace("Loading script file : [{}]", file);
+            }
+
+            CacheKeyT cacheKey = cacheKeyForFile(scriptNameExt.v1(), scriptNameExt.v2());
+            if (cacheKey == null) {
+                return;
```

Comment:
The implementations of `cacheKeyForFile` are responsible for logging or if it is appropriate. I can leave a comment about that. And update the javadocs for `cacheKeyForFile`. Or change it's signature if you think that'd help. The idea is that `cacheKeyForFile` is where the `TemplateService` ignores everything but mustache silently and `ScriptService` silently ignores mustache but otherwise is picky.

I'd intended to be able to change that behavior as time goes on, assuming we one day want to move templates out of `conf/scripts`.